doc: Add README for basic, basic_lorawan, freefall_lorawan apps (#195)
TheThingsIndustries/generic-node-se: Software/app/basic_lorawan/README.md

# basic\_lorawan

## Description

> basic\_lorawan serves as a basic template for LoRaWAN applications.

Reviewer: use backtick instead of escaping the underscore in the application name

# basic\_lorawan

## Description

`basic_lorawan` serves as a basic template for LoRaWAN applications.

## Configuration

### Default

> 1. Follow [this](https://www.thethingsindustries.com/docs/devices/adding-devices/) guide to add your device to TTN.

Reviewer 2: Change all occurences of TTN to The Things Stack. We use TTN only when talking about community. These are the official distribution names: https://www.thethingsindustries.com/docs/download/

2. Change `APPEUI`, `DEVEUI` and `APPKEY` in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h) to the values for you specific device on TTN.
3. Uncomment the REGION parameters in [conf/lorawan_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/lorawan_conf.h) and set `ACTIVE_REGION` in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h). These should correspond to the location that you use these devices for.

### Low power

1. Make sure that `LOW_POWER_DISABLE` is kept at 0 in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h).
2. Set `GNSE_ADVANCED_TRACER_ENALBE` to 0 in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h). This will make the device consume less power, but will also disable UART.

> Additionally, try to remember that:

Reviewer: Please address this as well

> - Setting the debugger will consume more power. The debugger is set in the macro `DEBUGGER_ON` in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h).

Reviewer 2: ```suggestion
- Setting the debugger will consume more power. The debugger is set in the macro `DEBUGGER_ON` in [`conf/app_conf.h`](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h).
```

> - Keeping peripherals, such as the sensors, off by controlling the load switches will reduce power consumption significantly.

Reviewer: @marnixcro please consider this improvement?!

> - The less transmissions over time, the lower the power consumption will be. This is set by `APP_TX_DUTYCYCLE` in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h).

Reviewer 2: ```suggestion
- The less transmissions over time, the lower the power consumption will be. This is set by `APP_TX_DUTYCYCLE` in [`lora_app.h`](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h).
```

### LoRa extra's

> Some extra setup configurations can be done to change the behaviour of the LoRaWAN capabilities. These include:

Reviewer 2: ```suggestion
Some extra setup configurations can be done to change the behaviour of the LoRaWAN capabilities. These include:

```

Reviewer 2: Maybe say "behaviour of the LoRaWAN features" instead of capabilities

> 1. Setting the activation method (OTAA or ABP) in `LORAWAN_DEFAULT_ACTIVATION_TYPE` in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h). OTAA is recommended as it is the most secure.

Reviewer 2: Wrap lora_app.h as `lora_app.h` and repeat for all occurrences

Reviewer 2: You can also link this: https://www.thethingsindustries.com/docs/devices/abp-vs-otaa/

> 2. The data rate can be set in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h). The default configuration uses the ADR. Should you want to set your preferred data rate, set `LORAWAN_ADR_STATE` to `LORAMAC_HANDLER_ADR_OFF` and set `LORAWAN_DEFAULT_DATA_RATE` to your preference. A list of the options per region are shown in [Region.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/lib/STM32WLxx_LoRaWAN/LoRaWAN/Mac/region/Region.h) in the STM32WLxx_LoRaWAN library.

Reviewer 2: ```suggestion
2. The data rate can be set in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h). The default configuration uses the ADR. If you want to set your preferred data rate, set `LORAWAN_ADR_STATE` to `LORAMAC_HANDLER_ADR_OFF` and set `LORAWAN_DEFAULT_DATA_RATE` to your preference. A list of the options per region are shown in [Region.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/lib/STM32WLxx_LoRaWAN/LoRaWAN/Mac/region/Region.h) in the STM32WLxx_LoRaWAN library.
```

> 3. `ACC_FF_LORA_PORT` can be changed in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h), which is used to configure to transmission port. The LoRaWAN keys mentioned in the default section can be altered here as well.

Reviewer 2: ```suggestion
3. `ACC_FF_LORA_PORT` can be changed in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h), which is used to configure the transmission port. The LoRaWAN keys mentioned in the default section can be altered here as well.
```

### Debugger

For debugging, the firmware has to support it first. The debugger is set in the macro `DEBUGGER_ON` in [conf/app_conf.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/conf/app_conf.h).

## Observation

> The standard payload is "**AA BB CC**", which will repeat periodically, with intervals set in `APP_TX_DUTYCYCLE` in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h). The TX/RX lines will show UART output which can be used for debugging. The default UART configuration is set to 115200/8-N-1. If nothing can be seen on the TTN console, then you could check from this debug output if:

Reviewer 2: ```suggestion
The standard payload is "**AA BB CC**", which will repeat periodically, with intervals set in `APP_TX_DUTYCYCLE` in [lora_app.h](https://github.com/TheThingsIndustries/generic-node-se/blob/develop/Software/app/basic_lorawan/lora_app.h). The TX/RX lines will show the UART output which can be used for debugging. The default UART configuration is set to 115200/8-N-1. If nothing can be seen on The Things Stack Console, then you could check from this debug output if:
```

> 1. The correct keys are used, which are printed at the start of the application.

Reviewer 2: ```suggestion
1. The correct keys are used, i.e. if they match the ones printed at the start of the application.
```

Reviewer: Also a good suggestion, please update

2. The device is connecting properly, as it will indicate if connections have failed or not. It should display something like this:
```
###### = JOINED = OTAA =====================

> ```

Reviewer 2: Add a blank line before and after these 3 lines

Reviewer 2: Same below

> 3. The device is transmitting anything, since each event is also logged via UART. It should display something similar to this:

Reviewer 2: ```suggestion
3. The device is transmitting anything, since each event is also logged via UART. It should display something similar to:
```

```
###### ========== MCPS-Confirm =============
###### U/L FRAME:0005 | PORT:2 | DR:5 | PWR:0 | MSG TYPE:UNCONFIRMED
60s069:TX on freq 868500000 Hz at DR 5
60s080:SEND REQUEST
```

> 4. Downlinks are detected if used. These only occur after being explicitly set in the TTN console or another application, and will only be transmitted after the transmission of an uplink by the device. The GNSE should, after receiving the downlink, display something similar to this:

Reviewer 2: ```suggestion
4. Downlinks are detected only if they are used. These only occur after being explicitly set in the The Things Stack Console or another application, and will only be transmitted after the transmission of an uplink. After receiving the downlink, the GNSE should display something similar to:
```
